fix: STX swap failures when needing to approve erc20

infiniteflower opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR fixes an issue with STX and ERC20 tokens that require approval failing on chain.

## **Related issues**

Fixes:

## **Manual testing steps**

Submit 8 STX Swaps with ERC20 tokens that need approval, all of them should succeed.

You will need to revoke allowances manually each time https://etherscan.io/tokenapprovalchecker, as there's a current bug where `Edit Limit` does not work and always submits an infinite approval.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

infiniteflower committed:
chore: wip for custom swaps stx handling

infiniteflower committed:
chore: wip use special stx path for swaps specifically

infiniteflower committed:
chore: wip add a todo

infiniteflower committed:
chore: get everything in place for submitSwapsSmartTransaction testing

infiniteflower committed:
chore: get swaps stx status modal skeleton working

infiniteflower committed:
feat: show smart transaction status page, but with custom swaps data

infiniteflower committed:
feat: get timer working

infiniteflower committed:
chore: remove unneeded condition

infiniteflower committed:
fix: navigation after create new swap and fix quote expiry closing the status modal

infiniteflower committed:
fix: modal animating when error screen appears

infiniteflower committed:
fix: use evm chain id selector

infiniteflower committed:
fix: select evm chain

infiniteflower committed:
chore: improve naming

infiniteflower committed:
fix: missing quote

infiniteflower committed:
chore: add tests

infiniteflower committed:
chore: update snapshots

infiniteflower committed:
chore: refactor test setup

infiniteflower committed:
fix: show swap approve and trade info in activity list

infiniteflower committed:
chore: add more tests

infiniteflower committed:
test: add tests for smart transactions selector for current chain

infiniteflower committed:
fix: broken tests

infiniteflower committed:
fix: lint errors

infiniteflower committed:
fix: lint errors

infiniteflower committed:
chore: add more tests

infiniteflower committed:
chore: update snapshots

infiniteflower committed:
chore: add more tests

infiniteflower committed:
chore: write more tests

infiniteflower committed:
chore: add more tests

infiniteflower committed:
chore: remove await from non promise

infiniteflower committed:
chore: remove unused code

infiniteflower committed:
chore: remove unneeded cancel fees, just use empty array for now

infiniteflower committed:
chore: add more tests

infiniteflower committed:
chore: add more tests

infiniteflower committed:
fix: lint issues

infiniteflower committed:
chore: update snapshots

infiniteflower committed:
chore: throw error if server error during stx submission

infiniteflower committed:
chore: use approval flow to open stx modal instead

infiniteflower committed:
chore: remove unneeded tests

infiniteflower committed:
chore: pass in entire selected quote into useSwapsSmartTransaction

infiniteflower committed:
chore: commit extra file for quote stx

infiniteflower committed:
fix: quote possibly undefined

infiniteflower committed:
fix: lint error

infiniteflower committed:
chore: update snapshots

infiniteflower committed:
chore: fix tests

infiniteflower committed:
fix: activity list not showing swaps stx info

infiniteflower committed:
chore: remove unused code

dan437 commented on app/components/UI/Swaps/QuotesView.js:
```diff
@@ -1063,7 +1051,7 @@ function SwapsQuotesView({
           ).toString(10),
         });
 
-        if (isHardwareAddress || shouldUseSmartTransaction) {
```
Why was this removed? ` || shouldUseSmartTransaction`

infiniteflower replied:
Because the flow used to be

`handleCompleteSwap` > `handleApprovalTransaction` > `handleSwapTransaction` 

and all of those txs would be routed through the STX hook

Now it's `handleCompleteSwap` > `submitSwapsSmartTransaction` and we don't route the Swaps STX through the STX hook

infiniteflower replied:
And correct me if I'm remembering wrong, but STX doesn't support HW wallets on Mobile yet right?

dan437 commented on app/components/UI/Swaps/QuotesView.js:
```diff
@@ -1112,27 +1099,71 @@ function SwapsQuotesView({
+        });
+      } catch (e) {
+        Logger.log(LOG_PREFIX, 'Failed to submit smart transaction', e);
         setIsHandlingSwap(false);
```
Ideally we would show a user some error notification if there was e.g. some server error, but that can be done in another ticket.

dan437 commented on app/components/UI/Swaps/utils/useSwapsSmartTransaction.ts:
```diff
@@ -0,0 +1,177 @@
+    // Returns e.g.: { uuid: 'dP23W7c2kt4FK9TmXOkz1UM2F20' }
+    return response.uuid;
+  } catch (error) {
+    console.error(error);
```
We should use `Logger.error` here instead.

infiniteflower replied:
Done in c7f6f0711e0e4f8a44509a7e235ab739280c9090

dan437 commented on app/components/UI/Swaps/utils/useSwapsSmartTransaction.ts:
```diff
@@ -0,0 +1,177 @@
+          gas: approvalGas,
+        },
+        smartTransactionFees: {
+          fees: smartTransactionFees.approvalTxFees.fees.map((fee) => ({
```
I see this mapping is used in another place as well, can you move it into a single function to avoid duplication? Also, what if you would just do `fees: smartTransactionFees.approvalTxFees.fees` and then convert it later in `submitSmartTransaction` or some other function it calls?

infiniteflower replied:
Done in 5677fb34135bb34fed259a303709502afdbc5aca

infiniteflower replied:
Actually done in f127456eb7024b32bc724674937e08d900683cb4

dan437 commented on app/components/UI/TransactionElement/utils.js:
```diff
@@ -691,8 +691,20 @@ function decodeSwapsTx(args) {
+  // If the tx was a swaps smart transaction, the swapsTransactions id is the stx.uuid, rather than tx.id
+  // We need use the tx.hash and look up the stx with the same hash
+  const smartTransaction =
+    Engine.context.SmartTransactionsController.state.smartTransactionsState.smartTransactions[
```
For unsupported chains `smartTransactions[chainId]` should be undefined and then you would get this error I believe: `Cannot read properties of undefined (reading 'find')`

infiniteflower replied:
Done in 45d4e24eaed70fd195a43ec052a03c58cc009ae3

dan437 commented:
Great job with this fix! I've done a high-level testing for a few most important test cases and it seems to be working.

There are a lot of changes in the QuotesView.js file. Many are just formatting, but since this PR is for one of the most important features in mobile, we need to make sure QA folks will test these changes carefully.

infiniteflower committed:
chore: use logger.error

infiniteflower committed:
chore: add a guard for undefined stx chainIds

infiniteflower committed:
chore: reduce duplication

infiniteflower committed:
chore: just move the fee conversion into submit stx

infiniteflower committed:
fix: fallback to normal tx if stx swap fails

infiniteflower committed:
chore: add comment

infiniteflower committed:
Revert "chore: add comment"

This reverts commit 3d9fb8095f69820f0deaeab2575bd77364b1fbaa.

infiniteflower committed:
Revert "fix: fallback to normal tx if stx swap fails"

This reverts commit 2668886b0f3311663eb148888b960c3447d95c93.

davibroc committed:
Toggles off STX (#13872)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

infiniteflower committed:
chore: update snapshots

vinnyhoward reviewed (approved):
Approved for UX code owner files

sethkfman reviewed (approved):
LGTM - approved E2E files
